tensorflow/core/framework/tensor.cc

```diff
@@ -754,8 +756,18 @@ Status Tensor::BitcastFrom(const Tensor& other, DataType dtype,
   shape_.set_data_type(dtype);
   if (buf_ != other.buf_) {
     UnrefIfNonNull(buf_);
-    buf_ = other.buf_;
-    RefIfNonNull(buf_);
+    if (port::kLittleEndian) {
+      buf_ = other.buf_;
+      RefIfNonNull(buf_);
+    } else {
+      Tensor ts_ = tensor::DeepCopy(other);
```

Comment:
Yes, I agree that the first case (bitcasting between types of the same size) does not need a deep copy on BE machines. However, I think the second case (bitcasting to/from (u)int8_t) may still need a deep copy when the type of the other end (source or destination) is not (u)int8_t. Because on BE machines, in general we need two rounds of byte swapping, one for the input type, the other for the output type. 

Please let me know what you think. Thanks @cantonios !

Reply:
Why would you need a deep copy if one of the types is a byte?

You would probably only do this in order to inspect (or potentially modify) individual bytes from the tensor.  Do we need to guarantee that the bytes stream itself would be little endian?  If we don't, it's faster and more consistent to re-use the buffer, with the caller needing to know that bytes are stored in native format.

Reply:
@cantonios Yes, I agree. But I think if we do not keep the data stream as little endian, test cases such as `//tensorflow/compiler/tests:cast_ops_test_cpu` which do the comparison might fail.

Reply:
Did you verify this?  I would still like to avoid byte-swapping when bit-casting to char, assuming a native byte ordering.

Reply:
Thanks @cantonios , I got your point, but If we apply the code changes in this PR to make the behaviours of different Bitcast implementations consistent on BE machines, the above mentioned 3 test cases (`//tensorflow/compiler/tests:cast_ops_test_cpu`, `//tensorflow/compiler/tests:cast_ops_test_cpu_mlir_bridge_test` and `//tensorflow/python/ops/ragged:ragged_bitcast_op_test`) could pass. I've verified this.

Also I've updated the code change to make the copy and swapping only happen on different type sizes on BE machines and reverted the code change in `tensorflow/c/c_api_test.cc`. Could you please take a look? Thank you very much!